iceoryx_hoofs/include/iceoryx_hoofs/posix_wrapper/thread.hpp

```diff
@@ -28,10 +33,58 @@ constexpr uint64_t MAX_THREAD_NAME_LENGTH = 15U;
 
 using ThreadName_t = cxx::string<MAX_THREAD_NAME_LENGTH>;
 
-void setThreadName(pthread_t thread, const ThreadName_t& name) noexcept;
-ThreadName_t getThreadName(pthread_t thread) noexcept;
+enum class ThreadError
+{
+    EMPTY_CALLABLE,
+    INSUFFICIENT_MEMORY,
+    INSUFFICIENT_PERMISSIONS,
+    INSUFFICIENT_RESOURCES,
+    INVALID_ATTRIBUTES,
+    UNDEFINED
+};
+
+class Thread
+{
+  public:
+    using callable_t = cxx::function<void()>;
+
+    Thread(const Thread&) = delete;
+    Thread& operator=(const Thread&) = delete;
+    Thread(Thread&& other) = delete;
+    Thread& operator=(Thread&& rhs) = delete;
+
+    ~Thread() noexcept;
+
+    // @todo set name during creation and remove method
+    void setThreadName(const ThreadName_t& name) noexcept;
+
+    ThreadName_t getThreadName() noexcept;
+
+    friend class ThreadBuilder;
+    friend class cxx::optional<Thread>;
+
+  private:
+    Thread() noexcept = default;
+
+    static ThreadError errnoToEnum(const int errnoValue) noexcept;
+
+    static void* startRoutine(void* callable);
+
+    pthread_t m_threadHandle;
+    callable_t m_callable;
```

Comment:
We should never support detached threads with a posix thread wrapper. The reason is that a posix wrapper ties the resource to its lifetime and we want to explicitly use RAII.

When we support detached threads then the lifetime of the thread is decoupled from the lifetime of the posix wrapper and this can introduce serious issues and race conditions as we have seen in the ara binding.
Then one can easily overlook that a thread is still running and one has to introduce some other mechanism to manage the lifetime.

If this is really required some day I would propose a thread builder for detached threads which simply returns an expected which states if this worked or not.